drivers/flash/soc_flash_ite_it8xxx2.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2020 ITE Corporation. All Rights Reserved.
+ * <jyunlin.chen@ite.com.tw>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/flash.h>
+#include <errno.h>
+#include <init.h>
+#include <soc.h>
+
+
+#define CONFIG_FLASH_SIZE	DT_REG_SIZE_BY_IDX(DT_NODELABEL(flash0), 0)
+#define FLASH_BLOCK_WRITE_UNIT_SIZE DT_PROP(DT_NODELABEL(flash0),\
+					erase_block_size)
+#define kStatus_Success     0
+
+struct flash_priv {
+	/*
+	 * HACK: flash write protection is managed in software.
+	 */
+	struct k_sem write_lock;
+	uint32_t pflash_block_base;
+};
+
+/**
+ * Interrupt vectors could be executed from flash hence the need for locking.
+ * The underlying MCUX driver takes care of copying the functions to SRAM.
+ *
+ * For more information, see the application note below on Read-While-Write
+ * http://cache.freescale.com/files/32bit/doc/app_note/AN4695.pdf
+ *
+ */
+
+static int flash_ite_erase(struct device *dev, off_t offset, size_t len)
+{
+	struct flash_priv *priv = dev->driver_data;
+	uint32_t addr;
+	int rc = -1;
+	unsigned int key;
+
+	if (k_sem_take(&priv->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	addr = offset + priv->pflash_block_base;
+
+	key = irq_lock();
+	irq_unlock(key);
+
+	k_sem_give(&priv->write_lock);
+
+	return (rc == kStatus_Success) ? 0 : -EINVAL;
+}
+
+static int flash_ite_read(struct device *dev, off_t offset,
+				void *data, size_t len)
+{
+	struct flash_priv *priv = dev->driver_data;
+	uint32_t addr;
+
+	/*
+	 * The MCUX supports different flash chips whose valid ranges are
+	 * hidden below the API: until the API export these ranges, we can not
+	 * do any generic validation
+	 */
+	addr = offset + priv->pflash_block_base;
+
+	memcpy(data, (void *) addr, len);
+
+	return 0;
+}
+
+static int flash_ite_write(struct device *dev, off_t offset,
+				const void *data, size_t len)
+{
+	struct flash_priv *priv = dev->driver_data;
+	uint32_t addr;
+	int rc = -1;
+	unsigned int key;
+
+	if (k_sem_take(&priv->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	addr = offset + priv->pflash_block_base;
+
+	key = irq_lock();
+	irq_unlock(key);
+
+	k_sem_give(&priv->write_lock);
+
+	return (rc == kStatus_Success) ? 0 : -EINVAL;
+}
+
+static int flash_ite_write_protection(struct device *dev, bool enable)
+{
+	struct flash_priv *priv = dev->driver_data;
+	int rc = 0;
+
+	if (enable) {
+		rc = k_sem_take(&priv->write_lock, K_FOREVER);
```

Comment:
It doesn't appear that this can return anything other than 0 when waiting forever. Might not be worth propagating an error that cannot happen (especially since disable case cannot error)

Reply:
Thanks you for reviewing.
This file has many details need to be modified.
And it cause build error at the latest version zephyr os.
So, We decide to remove it and will pull a new request after rewriting the flash driver.